### PR TITLE
refactor: define detailed error types

### PR DESCRIPTION
**Pull request by usamoi**
1. A clearer interface exposed by `service`.
2. RPC state machine is removed and generated by macros now.

A part of #307 and #349.

### CODE REVIEW

File: src/ipc/mod.rs [hunk 1]
```diff
@@ -1,59 +1,328 @@
...
+    socket: ServerSocket,
+}
+
+impl ServerRpcHandler {
+    pub(super) fn new(socket: ServerSocket) -> Self {
+        Self { socket }
+    }
+}
+
+macro_rules! define_packets {
```
Comment (cutecutecat) [inline src/ipc/mod.rs:102]: Long macro is really hard to understand. Do we have another way?

Comment (usamoi) [reply src/ipc/mod.rs:102]: #349

File: src/bgworker/mod.rs [hunk 2]
```diff
@@ -63,8 +72,6 @@ fn main() {
         if Worker::check(path.to_owned()) {
```
[the remark this answers is not in the document]
Comment (usamoi) [reply src/bgworker/mod.rs]: Okay. Done in new commit.

File: src/bgworker/mod.rs [hunk 3]
```diff
@@ -1,26 +1,35 @@
 pub mod normal;
-pub mod upgrade;
+
+use std::sync::atomic::{AtomicBool, Ordering};
+
+static NORMAL: AtomicBool = AtomicBool::new(false);
```
[the remark this answers is not in the document]
Comment (usamoi) [reply src/bgworker/mod.rs]: Done in new commit.

File: crates/service/src/worker/mod.rs [hunk 4]
```diff
@@ -80,15 +76,70 @@ impl Worker {
...
+    pub fn _drop(&self, handle: Handle) -> Result<(), DropError> {
         let mut protect = self.protect.lock();
         if protect.indexes.remove(&handle).is_some() {
             protect.maintain(&self.view);
+            Ok(())
+        } else {
+            Err(DropError::NotExist)
         }
     }
+    pub fn _flush(&self, handle: Handle) -> Result<(), FlushError> {
```
[the remark this answers is not in the document]
Comment (usamoi) [reply crates/service/src/worker/mod.rs:91]: Removing `_` will introduce conflicts, `create` RPC and `create` `Worker`, `drop` RPC and `Arc::drop`. Do we have better names?

File: crates/service/src/worker/mod.rs [hunk 5]
```diff
@@ -80,15 +76,70 @@ impl Worker {
...
+        let view = instance.view().ok_or(DeleteError::Upgrade)?;
+        view.delete(pointer)?;
+        Ok(())
+    }
+    pub fn _basic_view(&self, handle: Handle) -> Result<InstanceView, BasicError> {
+        let view = self.view();
+        let instance = view.get(handle).ok_or(BasicError::NotExist)?;
+        instance.view().ok_or(BasicError::Upgrade)
+    }
+    pub fn _vbase_view(&self, handle: Handle) -> Result<InstanceView, VbaseError> {
```
[the remark this answers is not in the document]
Comment (usamoi) [reply crates/service/src/worker/mod.rs:127]: I suggest not.

We set function only because if we do not, there are self referential structures.

**Comment by usamoi** [comment]
PTAL @cutecutecat

**Review by cutecutecat — APPROVED** [review]

**Pull request merged**